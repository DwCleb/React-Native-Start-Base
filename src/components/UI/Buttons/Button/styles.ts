import styled from 'styled-components/native';

import { Button } from 'react-native-paper';

import { Metrics, Colors } from '~/theme';
import { ThemeType } from '~/theme/theme';

export enum ButtonMode {
  TEXT = 'text',
  OUTLINED = 'outlined',
  CONTAINED = 'contained',
}

export interface ButtonProps {
  title: string;
  onPress(): void;
  accessibilityLabel?: string;
  icon?: string;
  color?: string;
  mode?: ButtonMode;
  loading?: boolean;
  disabled?: boolean;
  uppercase?: boolean;
  small?: boolean;
  distance?: number[];
  ratio?: number;
}

export type Props = ButtonProps & { theme: ThemeType };

function getMargin({ distance }: Props): string {
  if (!distance) {
    return `${10}px ${5}px ${10}px ${5}px`;
  } else {
    switch (distance.length || 0) {
      case 1:
        return `${distance[0]}px`;
      case 2:
        return `${distance[0]}px ${distance[1]}px`;
      case 3:
        return `${distance[0]}px ${distance[1]}px ${distance[2]}px`;
      case 4:
        return `${distance[0]}px ${distance[1]}px ${distance[2]}px ${distance[3]}px`;
      default:
        return '0px';
    }
  }
}

export const MyButton = styled(Button).attrs(
  ({
    icon,
    color = Colors.palette.primary.blue,
    mode = ButtonMode.CONTAINED,
    uppercase = false,
    loading,
    ratio = 0.85,
  }: ButtonProps) => ({
    icon,
    color,
    mode,
    uppercase,
    contentStyle: { height: 40, width: Metrics.screen.width * ratio },
    loading,
  })
)`
  margin: ${(props: Props): string => getMargin(props)};
  justify-content: center;
  align-items: center;
`;
