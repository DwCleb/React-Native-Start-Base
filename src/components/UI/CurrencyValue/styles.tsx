import Typography from '../Typography';
import styled from 'styled-components/native';
import { scale, ThemeContextType } from '~/theme';

interface WrapperProps {
  color?: string;
  opacity: string;
  valueWeight: string;
  isCurrency?: boolean;
}

type ThemedProps = ThemeContextType & WrapperProps;

function getTitleColor({ theme, color }: ThemedProps): string {
  return color ? theme.currency.title[color] : theme.monochromatic.white;
}

export const LineTroughContainer = styled.View`
  flex-direction: row;
  align-items: center;
  justify-content: center;
  position: relative;
  margin-right: ${scale(3)}px;
`;

export const PercentageContainer = styled.View`
  flex-direction: row;
  align-items: center;
  justify-content: center;
`;

export const Value = styled(Typography)`
  color: ${(props: ThemedProps): string => `${getTitleColor(props)}${props.opacity === 'half' ? '7F' : ''}`};
  ${({ valueWeight }: ThemedProps): string => valueWeight && `font-weight: ${valueWeight}`};
  letter-spacing: ${(props: ThemedProps): string => (props.isCurrency ? `${scale(1.5)}` : `${scale(1)}`)}px;
`;

export const LineThrough = styled.View`
  width: 100%;
  height: ${scale(2)}px;
  background-color: yellow;
  position: absolute;
  bottom: auto;
`;

function padding(prop: number[] = []): string {
  switch (prop.length || 0) {
    case 1:
      return `${scale(prop[0])}px`;
    case 2:
      return `${scale(prop[0])}px ${scale(prop[1])}px`;
    case 3:
      return `${scale(prop[0])}px ${scale(prop[1])}px ${scale(prop[2])}px`;
    case 4:
      return `${scale(prop[0])}px ${scale(prop[1])}px ${scale(prop[2])}px ${scale(prop[3])}px`;
    default:
      return '0px';
  }
}

export const Container = styled.View`
  flex-direction: ${({ row }: { row: string }): string => (row ? 'row' : 'column')};
  align-items: center;
  justify-content: flex-start;
  padding: ${({ distance }: { distance: [] }): string => padding(distance)};
`;
