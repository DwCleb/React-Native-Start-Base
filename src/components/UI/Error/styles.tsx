import styled from 'styled-components/native';
import { scale, ThemeContextType } from '~/theme';
import Typography, { TypographyType } from '~/components/UI/Typography';
import Modal from 'react-native-modal';

//Modal
export const ErrorModal = styled(Modal)`
  margin: 0px;
`;

//Parent Component
export const MainContainer = styled.SafeAreaView`
  flex: 1;
  width: 100%;
  background-color: ${({ theme }: ThemeContextType): string => theme.background};
`;

//Content's ScrollView
export const Scroll = styled.ScrollView.attrs(() => ({
  keyboardShouldPersistTaps: 'handled',
  showsVerticalScrollIndicator: false,
  contentContainerStyle: { flex: 1, paddingTop: scale(35) },
}))`
  height: auto;
`;

//Content Container
export const ContentContainer = styled.View`
  flex: 1;
  margin: ${scale(30)}px;
`;

//Page Title
export const Title = styled(Typography).attrs(() => ({
  type: TypographyType.H2,
  weight: 'regular',
  align: 'left',
}))`
  color: ${({ theme }: ThemeContextType): string => theme.text.black};
  margin-bottom: ${scale(10)}px;
  letter-spacing: ${scale(1)}px;
`;

export const SubTitle = styled(Typography).attrs(() => ({
  type: TypographyType.H5,
  weight: 'regular',
  align: 'left',
}))`
  color: ${({ theme }: ThemeContextType): string => theme.text.black};
  letter-spacing: ${scale(1)}px;
`;

//Error message
export const ErrorMessageContainer = styled.View`
  border-radius: ${scale(8)}px;
  height: ${scale(100)}px;
  padding: ${scale(20)}px ${scale(40)}px;
  justify-content: center;
  align-items: center;
  background-color: ${({ theme }: ThemeContextType): string => theme.error.card.background};
`;

export const ErrorMessage = styled(Typography).attrs(() => ({
  type: TypographyType.P2,
  weight: 'regular',
  align: 'center',
  numberOfLines: 2,
}))`
  color: ${({ theme }: ThemeContextType): string => theme.monochromatic.black};
  letter-spacing: ${scale(1)}px;
`;

//Buttons container
export const ButtonsContainer = styled.View`
  margin-bottom: ${scale(30)}px;
`;
