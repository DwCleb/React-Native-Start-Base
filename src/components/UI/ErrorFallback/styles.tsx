import styled from 'styled-components/native';
import { scale } from '~/theme';

import { Logo } from '~/assets';

export const ErrorImage = styled.Image.attrs(() => ({
  source: Logo,
  resizeMode: 'contain',
  align: 'center',
  marginTop: scale(30),
  marginBottom: scale(30),
  alignSelf: 'center',
}))`
  width: ${scale(82)}px;
`;
