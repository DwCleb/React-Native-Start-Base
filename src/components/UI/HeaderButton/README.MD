```jsx
// TODO: create a Wrapper to react-styleguidist to provide theme to all components shown
import ThemeManager from '../../../theme/ThemeContext';
import HeaderButton, { HeaderButtonEnum } from '.';

<ThemeManager>
  <HeaderButton action={() => {}} type={HeaderButtonEnum.Back} />
</ThemeManager>;
```
