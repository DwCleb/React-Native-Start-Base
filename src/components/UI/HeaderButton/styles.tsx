import styled from 'styled-components/native';

/** Theme */
import { scale, ThemeContextType } from '~/theme';

/** Assets */
import { BackIcon as BackIconAsset, CrossIcon as CrossIconAsset } from '~/assets';

/* Dependecies */
import { I18nManager } from 'react-native';

enum ButtonEnum {
  Back = 'back',
  Close = 'close',
}

interface IconProps {
  type: ButtonEnum;
}

export const ButtonContainer = styled.TouchableOpacity.attrs(() => ({
  hitSlop: { top: 30, left: 30, bottom: 30, right: 30 },
}))`
  margin-top: 0px;
  margin-left: ${(props: IconProps): string => (props.type === ButtonEnum.Back ? `${scale(20)}px` : '0px')};
  padding-right: ${(props: IconProps): string => (props.type === ButtonEnum.Back ? `${scale(30)}px` : '0px')};
`;

export const BackIcon = styled.Image.attrs((props: IconProps) => ({
  source: props.type === ButtonEnum.Back ? BackIconAsset : CrossIconAsset,
  resizeMode: 'contain',
}))`
  transform: scaleX(${I18nManager.isRTL ? -1 : 1});
  tint-color: ${(props: ThemeContextType): string => `${props.theme.navigation.back}`};
  height: ${(props: IconProps): string => (props.type === ButtonEnum.Back ? `${scale(16)}px` : `${scale(26)}px`)}
  width: ${(props: IconProps): string => (props.type === ButtonEnum.Back ? `${scale(9)}px` : `${scale(26)}px`)}
`;
