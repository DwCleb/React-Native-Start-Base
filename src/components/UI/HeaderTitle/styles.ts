import styled from 'styled-components/native';

/* Components */
import Typography, { TypographyType } from '../Typography';

/* Theme */
import { scale, ThemeContextType } from '~/theme';

export const Header = styled.View`
  padding-bottom: ${scale(35)}px;
`;

export const TitleTypography = styled(Typography).attrs({
  type: TypographyType.H3,
  align: 'left',
  distance: [0, 0, 0, 0],
})`
  align-self: flex-start;
  margin-left: ${scale(20)}px;
  margin-right: ${scale(20)}px;
  color: ${({ theme }: ThemeContextType): string => theme.headerTitle.text};
`;

export const SubtitleTypography = styled(Typography).attrs({
  type: TypographyType.P1,
  align: 'left',
  distance: [0, 0, 0, 0],
})`
  align-self: flex-start;
  margin-left: ${scale(20)}px;
  margin-right: ${scale(20)}px;
  color: ${({ theme }: ThemeContextType): string => theme.headerTitle.text};
`;
