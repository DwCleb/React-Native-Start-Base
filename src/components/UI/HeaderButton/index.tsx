import React from 'react';

import { ButtonContainer, BackIcon } from './styles';

export enum HeaderButtonEnum {
  Back = 'back',
  Close = 'close',
}

export interface HeaderButtonProps {
  action: () => void;
  type: HeaderButtonEnum;
}

export default function HeaderButton(props: HeaderButtonProps): JSX.Element {
  return (
    <ButtonContainer onPress={props.action} type={props.type}>
      <BackIcon type={props.type} />
    </ButtonContainer>
  );
}
