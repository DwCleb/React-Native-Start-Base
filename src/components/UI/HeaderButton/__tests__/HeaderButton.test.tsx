import React from 'react';

import { create, ReactTestRenderer, act } from 'react-test-renderer';

import { ThemeProvider } from 'styled-components/native';

import HeaderButton, { HeaderButtonProps, HeaderButtonEnum } from '../index';

import { ButtonContainer } from '../styles';

import darkTheme from '~/theme/ThemeContext/dark';

const theme = darkTheme.theme;

const action = jest.fn();

const buttonProps = {
  action,
  type: HeaderButtonEnum.Back,
};

const componentToTest = (props: HeaderButtonProps): JSX.Element => (
  <ThemeProvider theme={theme}>
    <HeaderButton {...props} />
  </ThemeProvider>
);

describe('<HeaderButton /> component', () => {
  const wrapper: ReactTestRenderer = create(componentToTest(buttonProps));
  const root: ReactTestRenderer['root'] = wrapper.root;

  beforeAll(() => {
    jest.useFakeTimers();
  });

  it('Should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('Should call function passed through props when pressed', () => {
    const button = root.findAllByType(ButtonContainer)[0];
    act(() => {
      button.props.onPress();
    });
    expect(action).toHaveBeenCalled();
  });

  it('It should render the right icon when type=close', () => {
    jest.mock('react-native/Libraries/ReactNative/I18nManager', () => ({
      isRTL: true,
    }));

    act(() => {
      wrapper.update(componentToTest({ ...buttonProps, type: HeaderButtonEnum.Close }));
    });

    expect(wrapper.root.findAllByType(ButtonContainer)[0].props.type).toEqual(HeaderButtonEnum.Close);
  });
});
