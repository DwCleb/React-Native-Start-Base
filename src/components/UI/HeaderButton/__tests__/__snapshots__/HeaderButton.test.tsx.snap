// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HeaderButton /> component Should render correctly 1`] = `
<View
  accessible={true}
  focusable={true}
  hitSlop={
    Object {
      "bottom": 30,
      "left": 30,
      "right": 30,
      "top": 30,
    }
  }
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "marginLeft": 40,
      "marginTop": 0,
      "opacity": 1,
      "paddingRight": 60,
    }
  }
>
  <Image
    resizeMode="contain"
    source={
      Object {
        "testUri": "../../../src/assets/icons/back.png",
      }
    }
    style={
      Array [
        Object {
          "height": 32,
          "tintColor": "#FFFFFF",
          "transform": Array [
            Object {
              "scaleX": 1,
            },
          ],
          "width": 18,
        },
      ]
    }
    type="back"
  />
</View>
`;
